File: app/src/main/java/nz/scuttlebutt/tremola/ssb/peering/discovery/LookUpUDP.java
package nz.scuttlebutt.tremola.ssb.peering.discovery;


import android.content.Context;
import android.util.Log;
import nz.scuttlebutt.tremola.MainActivity;
import nz.scuttlebutt.tremola.ssb.core.SSBid;
import nz.scuttlebutt.tremola.utils.Constants;

import java.io.IOException;
import java.net.*;
import java.util.Objects;
import java.util.concurrent.locks.ReentrantLock;

public class LookUpUDP extends LookUpClient {

    private String broadcastAddress;

    public LookUpUDP(LookUp lookUp, Context context, int port, SSBid ed25519KeyPair) {
        super(lookUp, context, port, ed25519KeyPair);
    }

    public void prepareQuery(String broadcastAddress) {
        this.broadcastAddress = broadcastAddress;
    }

    @Override
    public void sendQuery(String broadcastMessage) {
        try {
            final DatagramSocket datagramSocket = new DatagramSocket();
            datagramSocket.setBroadcast(true);
            InetAddress receiverAddress = InetAddress.getByName(broadcastAddress);

            final DatagramPacket datagramPacket = new DatagramPacket(
                    broadcastMessage.getBytes(),
                    broadcastMessage.length(),
                    receiverAddress,
                    port
            );

            Log.d("Send Packet", new String(datagramPacket.getData()));

            datagramSocket.send(datagramPacket);
            datagramSocket.close();
        } catch (IOException e) {
            Log.e("SendMessage", e.getMessage());
            e.printStackTrace();
        }
    }


    public void listen(ReentrantLock lock) throws InterruptedException {
        byte[] buf = new byte[512];
        DatagramPacket ingram = new DatagramPacket(buf, buf.length);
        while (true) {
            try {
                Objects.requireNonNull(((MainActivity) context).getLookup_socket()).receive(ingram);
            } catch (Exception e) {
                synchronized (lock) {
                    lock.wait(Constants.Companion.getLOOKUP_INTERVAL());
                }
                continue;
            }
            String incoming = new String(ingram.getData(), 0, ingram.getLength());
            for (String i : incoming.split(";")) {
                Log.e("lu_rx " + ingram.getLength(), "<" + i + ">");
                if (i.startsWith("{\"targetName")) {
                    lookUp.acceptQuery(i);
                    lookUp.processQuery();
                } else if (i.startsWith("{\"targetId")) {
                    lookUp.acceptReply(i);
                    lookUp.processReply();
                }
            }
        }
    }
}

File: app/src/main/java/nz/scuttlebutt/tremola/ssb/peering/discovery/LookupUDP.kt
package nz.scuttlebutt.tremola.ssb.peering.discovery

import android.content.Context
import android.util.Log
import nz.scuttlebutt.tremola.MainActivity
import nz.scuttlebutt.tremola.ssb.core.SSBid
import nz.scuttlebutt.tremola.utils.Constants.Companion.LOOKUP_INTERVAL
import java.io.IOException
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.ReentrantLock

class LookupUDP(
    lookup: Lookup,
    context: Context,
    ed25519KeyPair: SSBid,
    lock: ReentrantLock,
    private val port: Int,
    private val broadcastAddress: String
) :
    LookupClient(lookup, context, ed25519KeyPair, lock) {
    private var datagramSocket: DatagramSocket? = null

    init {
        try {
            datagramSocket = DatagramSocket()
            datagramSocket!!.broadcast = true
        } catch (e: Exception) {
            active = false
        }
    }

    override fun sendQuery(broadcastMessage: String) {
        try {
            if (datagramSocket!!.isClosed) {
                datagramSocket = DatagramSocket()
                datagramSocket!!.broadcast = true
            }
            val receiverAddress = InetAddress.getByName(broadcastAddress)
            val datagramPacket = DatagramPacket(
                broadcastMessage.toByteArray(),
                broadcastMessage.length,
                receiverAddress,
                port
            )
            datagramSocket!!.send(datagramPacket)
            datagramSocket!!.close()
            Log.e("lu_wr ", "<${String(datagramPacket.data)}>")
        } catch (e: IOException) {
            Log.e("SEND_ERROR", e.message!!)
            e.printStackTrace()
        }
    }

    override fun run() {
        val buf = ByteArray(512)
        val ingram = DatagramPacket(buf, buf.size)
        active = true
        while (active) {
            try {
                (context as MainActivity).lookup_socket!!.receive(ingram)
            } catch (e: Exception) {
                synchronized(lock) {
                    try {
                        lock.tryLock(LOOKUP_INTERVAL, TimeUnit.MILLISECONDS)
                    } catch (ex: InterruptedException) {
                        Log.e("UDP LOCK", ex.message!!)
                    }
                }
                continue
            }
            val incoming = String(ingram.data, 0, ingram.length)
            for (i in incoming.split(";").toTypedArray()) {
                Log.e("lu_rx " + ingram.length, "<$i>")
                if (i.contains("\"msa\"")) {
                    lookup.processQuery(i)
                } else if (i.contains("\"targetId")) {
                    lookup.processReply(i)
                }
            }
        }
    }
}